### DISCUSSION TITLE
How to deploy Nuxt 3 app to CPanel VPS Server?

### DISCUSSION THREAD

**Discussion by murugappanrm** [nuxt/nuxt#16340] [Questions]
We are large group of newbies to Nuxt3 and we love it. But all that happiness went down the tubes as quick as our excitement. We have looked every where on how to depoy a Nuxt 3 app to the CPANEL VPS server. Its all good and wonderful that all the techie code shown by super developer but is hopeless to us who know nothing server setups, networking and domains. We know basic stuff on these matters. 

Our disappointment is that Nuxtjs is simple to develop but is pain when comes to deploy. We are so used to zip-upload-extract-run process in using other platforms like Codeigniter, pure Web Stack or vuejs. We need help on this with Nuxt3 app. The consensus is that if Nuxtjs is so complicated and troublesome, we want to dump and de-list in our strategy and go back to the safer model. 

Can anyone help in providing a simple step by step guide (without all that fiddling with node and server backend - remember we using VPS shared hosting) on this so that by simply executing the steps we should be able to run the app? 

All help much appreciated. Thank you in advance.

**Comment by TheAlexLichter** [comment]
Disclaimer: I have little experience with CPanel, Plesk and whatnot but it still might be worth a shot.

1. Run the build command as usual (preset should be `node-server` and is set to that by default)
2. Take the whole `.output` folder and put it on CPanel (e.g. in a `nuxt-app` root)
3. Set up a Node.js app, choose the lates LTS Node version and point the starting file to `.output/server/index.mjs`

That should do the trick

**Reply by murugappanrm** [reply]
Hi @manniL 

Thank you for the help. I will try you suggestion and will update here.

**Comment by murugappanrm** [comment]
@manniL 

Seriously, Nuxt is serious disappointment to us in ASEAN region developer community. The deployment processes and entire documentation is shamefully ineffective for people like us who want to migrate from other platforms. There is not even a simple section on deployment to CPanel on Centos. 

Many of use our own local Shared Hosting Providers like Hostinger, InMotion, Exabyte etc. due to constrains by local authorities. I have tried all sorts of tricks and nothing works. Are the developers of Nuxt telling us to go find other solutions?

**Reply by TheAlexLichter** [reply]
I won't follow up here, instead, read https://github.com/nuxt/framework/discussions/8742 please.

**Reply by drobertson-dev** [reply]
Deploying to a VPS with CPanel or Plesk as a control panel is pretty simple. It's just like the answer above.  If you are using a shared provider, they probably gave you access to a control panel.  You need to ensure they support a node environment because many managed hosting doesn't support node; they focus on WordPress or their proprietary website solution.  You can deploy to Cloudflare workers or pages for free, and the deployment is literally seconds, and your site is up. If you have access to the internet, you have access to Cloudflare.

**Reply by murugappanrm** [reply]
@daver you do not understand the need. Servers like Cloudflare are located outside of our countries. Here in the ASEAN regions locating applications and data outside the countries are prohibited to prevent idiotic abuse like Facebook. So we need to deploy to local shared servers like Exabyte, InmMotion (who have a local base as well). 

I hope this makes it clear why we do it and especially our data goes into terabytes.

**Comment by drobertson-dev** [comment]
If the shared hosting supports node, then you can deploy it easily. It’s basically copy and paste. Like the above comment says. If the shared hosting does not support Node then you are out of luck until there is an alternative hosting solution.

On Nov 13, 2022, at 21:57, Murugappan Ramanathan ***@***.***> wrote:

﻿

@daver<https://github.com/daver> you do not understand the need. Servers like Cloudflare are located outside of our countries. Here in the ASEAN regions locating applications and data outside the countries are prohibited to prevent idiotic abuse like Facebook. So we need to deploy to local shared servers like Exabyte, InmMotion (who have a local base as well).

I hope this makes it clear why we do it and especially our data goes into terabytes.

**Comment by murugappanrm** [comment]
Hi @daver987 

Allow me to give you the scenario:
(1) I need to run the app as **"sysdev.emastpa.com.my/nuxtapp".**
(2) To do that the app need to be loaded in a **"public_html/nuxtapp"** directory. The "public_html" is similar to the www root directory under Apache.

Our shared VPS has the latest nodejs installed.  I copied and pasted the whole repo into a directory "nuxtapp-dev" folder in the server. I could run the npm dev and then access the site using **"sysdev.emastpa.com.my:3000/"**. The app loads wonderfully. I could do a build too. But the app is generated into a **.output** directory. That is all the deployment documentation is good for. After that, i do not know what to do and there is no help in the documentation. 

How do i get the contents of .output to able to load the app as (1) above? That is able to run the app from "public_html/nuxtapp" folder. Appreciate all the help.

**Reply by drobertson-dev** [reply]
It sounds like you are nearly there, You can change the name of the .output folder, you can also tell node which folder to serve from. This sounds more like a node issue than a Nuxt 3 issue.  I set up a few sites on private  a VPS, all of them were Nuxt 2, which is more of bear to get set up, and I found it to be quite brittle.  I would check the Node documentation, I think that's where you would possibly get more information.  I don't know what the security risks are like building on server etc. I think what you're asking for is actually out of the scope of what the Nuxt documentation should provide.  I will do an experiment myself and see, but I use Plesk and not Cpanel. But I also know you can't just switch from one to the other easily because of how integrated the actual server part is with the control panel.   If I get some time maybe I'll write something detailed for the documentation. This is the beauty of open source, it a community effort.

**Reply by murugappanrm** [reply]
@daver987 , any help to get this working will be fantastic. All i wish is that i need to able to load my app using the url **"sysdev.emastpa.com.my/nuxtapp**" without assigning any port number in the URL (if possible).  I do not want to do the dev or build on the server. As you rightly said, it would be a disaster in terms of security and control. I really love what nuxt 3 can do and its flexibility. The sad part is that i not familiar with Nodejs and i would not want to avoid too much learning curve. I am business solution developer and cannot afford the time to dive deep into these software. I assure you that your solution will go a long way with hundreds of solution developers in our region.

Looking forward to your help.
